# Digicamp 2021
### Introduction to Python for data analysis
www.digicamp.ch



### Pandas, Working with Dataset -- 1
Take away:
- Import a csv dataset in Python
- Relational operators and conditions
- Filtering and subsetting dataset

#### 22-07-2021
Mattia Biancaterra, mattia.biancaterra@usi.ch

#### Dataset provider
https://www.kaggle.com/datasets

Used in this section:
- Ad Click Prediction, https://www.kaggle.com/jahnveenarang/cvdcvd-vd

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Social_Network_Ads.csv')

In [3]:
df

User ID  Gender  Age  EstimatedSalary  Purchased
0    15624510    Male   19            19000          0
1    15810944    Male   35            20000          0
2    15668575  Female   26            43000          0
3    15603246  Female   27            57000          0
4    15804002    Male   19            76000          0
..        ...     ...  ...              ...        ...
395  15691863  Female   46            41000          1
396  15706071    Male   51            23000          1
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

[400 rows x 5 columns]

In [4]:
df.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

The computer has done something that is correct but makes no sense. Who can tell me what it is?

In [5]:
df.head(3)

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0

In [6]:
df.tail(3)

User ID  Gender  Age  EstimatedSalary  Purchased
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

In [7]:
#How to select colums?
df['User ID']
#to that I can append any method

0      15624510
1      15810944
2      15668575
3      15603246
4      15804002
         ...   
395    15691863
396    15706071
397    15654296
398    15755018
399    15594041
Name: User ID, Length: 400, dtype: int64

In [8]:
df.iloc[1]

User ID            15810944
Gender                 Male
Age                      35
EstimatedSalary       20000
Purchased                 0
Name: 1, dtype: object

In [9]:
df.iloc[[1, 67, 96]]

User ID  Gender  Age  EstimatedSalary  Purchased
1   15810944    Male   35            20000          0
67  15654574  Female   23            82000          0
96  15710257  Female   35            25000          0

#### What is a list?

In [10]:
my_list = [1,3,4,5]

In [11]:
my_list

[1, 3, 4, 5]

In [12]:
word_list = ['Age', 'Gender'] 

In [13]:
word_list

['Age', 'Gender']

Can I modify the dataset? yes sure! Add a new columns

In [14]:
df['New Column'] = df['Age'] + 4

In [15]:
df

User ID  Gender  Age  EstimatedSalary  Purchased  New Column
0    15624510    Male   19            19000          0          23
1    15810944    Male   35            20000          0          39
2    15668575  Female   26            43000          0          30
3    15603246  Female   27            57000          0          31
4    15804002    Male   19            76000          0          23
..        ...     ...  ...              ...        ...         ...
395  15691863  Female   46            41000          1          50
396  15706071    Male   51            23000          1          55
397  15654296  Female   50            20000          1          54
398  15755018    Male   36            33000          0          40
399  15594041  Female   49            36000          1          53

[400 rows x 6 columns]

### Filtering
A subset is a part of a dataset that is choosen by some criteria. 

Before starting what are relational operators? 

$$\geq \hspace{0.4cm} \leq \hspace{0.4cm} < \hspace{0.4cm} >  \hspace{0.4cm}== \hspace{0.4cm}!=$$

In [16]:
# error example
3=4

SyntaxError: cannot assign to literal (<ipython-input-16-fa6227cda19a>, line 2)

In [17]:
3==4

False

In [18]:
3!=4

True

In [19]:
3>=4

False

In [20]:
3<=4

True

In [21]:
df['Gender'] == 'Male'

0       True
1       True
2      False
3      False
4       True
       ...  
395    False
396     True
397    False
398     True
399    False
Name: Gender, Length: 400, dtype: bool

Ok, but now that I have a list of True and False, what I can do? 

True has value 1 and False hase value 0. 

In [22]:
list_ = [True, False, True, False, True, True]
sum_ = sum(list_)
sum_

4

In [23]:
sum(df['Gender'] == 'Male')

196

In [24]:
num_male = sum(df['Gender'] == 'Male')
num_male

196

In [25]:
# you can easily choose != instead of changing the name of the columns
num_female = sum(df['Gender'] != 'Male')
num_female = sum(df['Gender'] == 'Female')

In [26]:
pct_male = num_male/(num_male+num_female)
pct_male

0.49

In [27]:
pct_male = num_male/len(df)
pct_male

0.49

In [28]:
df['EstimatedSalary'][df['Age'] >= 30]

1       20000
7      150000
9       65000
13      18000
16      25000
        ...  
395     41000
396     23000
397     20000
398     33000
399     36000
Name: EstimatedSalary, Length: 300, dtype: int64

In [29]:
#Imagine you need to calculate the average salary of people aged 30 or more. How to subset with 2 conditions?
df['EstimatedSalary'][df['Age'] >= 30].mean()
# This litterally means give me the mean of the column estimated salary using only value in row where the value of the column age is above 30

73080.0

A quick note on ```.mean```. Try to call this method without telling what columns. For example call the mean of ```df[df['Gender'] == 'Male]```. Automatically the code will return the mean of all columns. This will be handy later.

In [30]:
df[df['Gender'] == 'Male'].mean()

User ID            1.568970e+07
Age                3.686735e+01
EstimatedSalary    6.764286e+04
Purchased          3.367347e-01
New Column         4.086735e+01
dtype: float64

Last but not least important, **multiple conditions**

$$ \& = and \hspace{1cm}| =  or $$

There is no limit to the number of conditions, but keep in mind that the expression must have logical sense.

In [31]:
multiple = (df['Age'] >=30) & (df['Gender'] == 'Male')
multiple

0      False
1       True
2      False
3      False
4      False
       ...  
395    False
396     True
397    False
398     True
399    False
Length: 400, dtype: bool

### Excercise

- Find the percentage of Users who have an estimated salary above 45'000.
- Find how many males, who have an estimated salary above 45'000, have made a purchase.
- Calculate the average ages of both males and females and store them in 2 new variable (avg_males, avg_females). 
- Now create a new column in the dataset called 'diff' that contains the difference between the ages and the average age of all the dataset. (to check: the average should be 37.655, and the first row should contain the result of 19 - 37.655)

### Solution

In [32]:
pct_user = sum(df['EstimatedSalary']>= 45000)/len(df)
pct_user

0.735

In [33]:
male_above = sum((df['Gender'] == 'Male') & (df['EstimatedSalary']>= 45000) & df['Purchased'] == 1)
male_above

47

In [34]:
avg_males = df['Age'][df['Gender'] == 'Male'].mean()
avg_females = df['Age'][df['Gender'] != 'Male'].mean()

In [35]:
df['diff'] = df['Age'] - df['Age'].mean()
df

User ID  Gender  Age  EstimatedSalary  Purchased  New Column    diff
0    15624510    Male   19            19000          0          23 -18.655
1    15810944    Male   35            20000          0          39  -2.655
2    15668575  Female   26            43000          0          30 -11.655
3    15603246  Female   27            57000          0          31 -10.655
4    15804002    Male   19            76000          0          23 -18.655
..        ...     ...  ...              ...        ...         ...     ...
395  15691863  Female   46            41000          1          50   8.345
396  15706071    Male   51            23000          1          55  13.345
397  15654296  Female   50            20000          1          54  12.345
398  15755018    Male   36            33000          0          40  -1.655
399  15594041  Female   49            36000          1          53  11.345

[400 rows x 7 columns]

A quick note before going ahead.

In [36]:
df.to_csv('my_new_df.csv')